### PR TITLE
Handle request cancellations

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

- `DasCustodySync`: cancel sync requests which are not appeared missing anymore
- Handle requests cancel in `RecoveringSidecarRetriever`: cancel recovery if all block requests are cancelled
- Add request cancel test cases

### CODE REVIEW

File: ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/retriever/DelayedDataColumnSidecarRetriever.java [hunk 1]
```diff
@@ -37,7 +38,16 @@ public void setDelay(Duration delay) {
   }
 
   private <T> SafeFuture<T> delay(Supplier<SafeFuture<T>> futSupplier) {
-    return asyncRunner.getDelayedFuture(delay).thenCompose(__ -> futSupplier.get());
+    com.google.common.base.Supplier<SafeFuture<T>> memoized = Suppliers.memoize(futSupplier::get);
```
Comment (zilm13) [inline ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/retriever/DelayedDataColumnSidecarRetriever.java]: Could use `java.util.function.Supplier`?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodySync.java [hunk 2]
```diff
@@ -100,27 +105,30 @@ private void fillUpIfNeeded() {
...
-                addPendingRequest(missingColumn);
-              }
+    final SafeFuture<Set<DataColumnSlotAndIdentifier>> missingColumnsFuture =
+        custody.retrieveMissingColumns().limit(maxPendingColumnRequests).collect(new HashSet<>());
 
-              if (missingColumnsToRequest.size() < newRequestCount) {
+    missingColumnsFuture
+        .thenAccept(
+            missingColumns -> {
+              if (missingColumns.size() < maxPendingColumnRequests) {
```
Comment (zilm13) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodySync.java:114]: Could you please add comment here for clarification: something like "we don't need to get new missing columns until new slot, we have got all"

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java [hunk 3]
```diff
@@ -206,4 +206,29 @@ void selectingBestPeerShouldRespectPeerMetrics() {
...
+
+    advanceTimeGradually(retrieverRound);
+    assertThat(custodyPeer.getRequests()).hasSize(1);
+    advanceTimeGradually(retrieverRound);
+    assertThat(custodyPeer.getRequests()).hasSize(2);
+
+    resp0_0.cancel(true);
+
+    advanceTimeGradually(retrieverRound);
+    assertThat(custodyPeer.getRequests()).hasSize(2);
```
Comment (zilm13) [inline ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java:231]: so it should grow to 3 if not cancelled and doesn't? Could you please add a comment

**Review by zilm13 — APPROVED** [review]
LGTM just some nits and please fix what CI doesn't like

File: ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/retriever/DelayedDataColumnSidecarRetriever.java (hunk 1 above)
Comment (Nashatyrev) [reply ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/retriever/DelayedDataColumnSidecarRetriever.java]: Used `CancelableFuture` here. No more relevant

**Pull request merged**